fdbserver/workloads/FailoverWithSSLag.actor.cpp

```diff
@@ -116,56 +116,90 @@ struct FailoverWithSSLagWorkload : TestWorkload {
 		return true;
 	}
 
-	ACTOR static Future<Optional<Version>> fetchDatacenterLag(FailoverWithSSLagWorkload* self, Database cx) {
+	// Fetches details (versions and seconds) of the specified type of lag (tlog/storage server/data center lag) from
+	// the given status json document.
+	bool fetchLagFromStatusObject(std::string path, StatusObjectReader& statusObj, Version& versions, double& seconds) {
+		StatusObjectReader lagObject;
+		if (!statusObj.get(path, lagObject)) {
+			return false;
+		}
+
+		if (!lagObject.get("versions", versions)) {
+			return false;
+		}
+
+		if (!lagObject.get("seconds", seconds)) {
+			return false;
+		}
+
+		return true;
+	}
+
+	ACTOR static Future<Optional<Version>> fetchStorageServerLag(FailoverWithSSLagWorkload* self, Database cx) {
+		state double startTime = now();
 		StatusObject result = wait(StatusClient::statusFetcher(cx));
+		double duration = now() - startTime;
+
 		StatusObjectReader statusObj(result);
 		StatusObjectReader statusObjCluster;
 		if (!statusObj.get("cluster", statusObjCluster)) {
-			TraceEvent("DcLagNoCluster");
+			TraceEvent("SSLagNoCluster");
 			return Optional<Version>();
 		}
 
-		StatusObjectReader dcLag;
-		if (!statusObjCluster.get("datacenter_lag", dcLag)) {
-			TraceEvent("DcLagNoLagData");
+		// Fetch the lag between primary and remote tlogs.
+		Version tlogLagInVersions = 0;
+		double tlogLagInSeconds = 0;
+		if (!self->fetchLagFromStatusObject("logserver_lag", statusObjCluster, tlogLagInVersions, tlogLagInSeconds)) {
+			TraceEvent("NoLogServerLagData");
 			return Optional<Version>();
 		}
 
-		Version versions = 0;
-		double seconds = 0;
-		if (!dcLag.get("versions", versions)) {
-			TraceEvent("DcLagNoVersions");
+		// Fetch the lag between primary and remote storage servers.
+		Version ssLagInVersions = 0;
+		double ssLagInSeconds = 0;
+		if (!self->fetchLagFromStatusObject("storageserver_lag", statusObjCluster, ssLagInVersions, ssLagInSeconds)) {
+			TraceEvent("NoStorageServerLagData");
 			return Optional<Version>();
 		}
-		if (!dcLag.get("seconds", seconds)) {
-			TraceEvent("DcLagNoSeconds");
+
+		// Fetch the lag between primary and remote data centers.
+		Version dcLagInVersions = 0;
+		double dcLagInSeconds = 0;
+		if (!self->fetchLagFromStatusObject("datacenter_lag", statusObjCluster, dcLagInVersions, dcLagInSeconds)) {
+			TraceEvent("NoDataCenterLagData");
 			return Optional<Version>();
 		}
-		TraceEvent("DcLag").detail("Versions", versions).detail("Seconds", seconds);
-		return versions;
+
+		TraceEvent("LagInfo")
+		    .detail("LogServerLagInVersions", tlogLagInVersions)
+		    .detail("LogServerLagInSeconds", tlogLagInSeconds)
+		    .detail("StorageServerLagInVersions", ssLagInVersions)
+		    .detail("StorageServerLagInSeconds", ssLagInSeconds)
+		    .detail("DataCenterLagInVersions", dcLagInVersions)
+		    .detail("DataCenterLagInSeconds", dcLagInSeconds)
+		    .detail("TimeToFetchStatus", duration);
+
```

Comment:
I notice that the tlog lag is also going up along with the storage server lag (even though the log router - tlog connections are not clogged):

> LogServerLagInVersions="9705655" StorageServerLagInVersions="6817704"
> LogServerLagInVersions="16274637" StorageServerLagInVersions="13372306"
> LogServerLagInVersions="21356690" StorageServerLagInVersions="18686031"
> LogServerLagInVersions="29863991" StorageServerLagInVersions="27003746" 
> 

I think it is because the log router stops fetching versions if the remote storage servers don't pop versions from remote logs (though I haven't verified if this is what is actually happening here).

Reply:
> I think it is because the log router stops fetching versions if the remote storage servers don't pop versions from remote logs

It appears that's what is going on here. I don't see the version fetched by the log router(s) (and also the "MinPopped" version) changing in logRouter metrics:

> Time="166.719308" Type="LogRouterMetrics" Version="453139404" MinPopped="448139404" FetchedVersions="5000000"
> ....
> Time="181.720549" Type="LogRouterMetrics" Version="453139404" MinPopped="448139404" FetchedVersions="5000000"

Reply:
Yes. The Log routers keep at most 5M version in memory. If remote tlogs are not popping, then log routers stop fetching data. I guess what happened in the real cluster is that SS can't keep up with writing data to disks faster enough, so there was SS lag, but not tlog lag.